webpack/scenes/ContentViews/Table/ContentViewTable.js

```diff
@@ -0,0 +1,140 @@
+import React, { useState, useEffect } from 'react';
+import {
+  Table,
+  TableHeader,
+  TableBody,
+} from '@patternfly/react-table';
+import PropTypes from 'prop-types';
+
+import Loading from './Loading';
+import tableDataGenerator, { buildColumns } from './tableDataGenerator';
+import emptyRows from './emptyRows';
+import './ContentViewTable.scss';
+
+const contentViewShowFakeData = require('../data/show');
+
+const ContentViewTable = ({ contentViews }) => {
+  const [detailsMap, setDetailsMap] = useState({}); // Map of CV id to details object
+  const [table, setTable] = useState({ rows: [], columns: [] });
+  // Map of CV id to expanded cell, if id not present, row is not expanded
+  const [expandedColumnMap, setExpandedColumnMap] = useState({});
+  const cvsPresent = contentViews && contentViews.results && contentViews.results.length > 0;
+
+  useEffect(
+    () => {
+      if (!contentViews) return;
+      if (cvsPresent) {
+        const tableData = tableDataGenerator(
+          contentViews,
+          detailsMap,
+          expandedColumnMap,
+        );
+        setTable(tableData);
+      } else {
+        setTable({ columns: buildColumns(), rows: emptyRows });
+      }
+    },
+    [contentViews, detailsMap, expandedColumnMap],
+  );
+
+  const cvIdFromRow = ({ details: { props: rowProps } }) => rowProps.contentviewid;
+
+  const loadDetails = (id) => {
+    if (detailsMap[id]) return;
+    // Replace with API call
+    setTimeout(() => {
+      const details = contentViewShowFakeData;
+      setDetailsMap(prev => ({ ...prev, [id]: details }));
+    }, 300);
+  };
+
+  const onSelect = (event, isSelected, rowId) => {
+    let rows;
+    if (rowId === -1) {
+      rows = table.rows.map((row) => {
+        row.selected = isSelected;
+        return row;
+      });
+    } else {
+      rows = [...table.rows];
+      rows[rowId].selected = isSelected;
+    }
+
+    setTable(prevTable => ({ ...prevTable, rows }));
+  };
+
+  const onExpand = (_event, _rowIndex, colIndex, isOpen, rowData, _extraData) => {
+    const { rows } = table;
+    const contentViewId = cvIdFromRow(rowData);
+    // adjust for the selection checkbox cell being counted in the index
+    const adjustedColIndex = colIndex - 1;
+
+    if (!isOpen) {
+      setExpandedColumnMap(prev => ({ ...prev, [contentViewId]: adjustedColIndex }));
```

Comment:
I agree and also wish that the table wasn't one giant object because you have to rebuild it each time a row is expanded. 

I'll brainstorm other options, but if we are using pf4, we shouldn't try to make it something it is not, and use it as expected. If you have a good suggestion other than the mapping I can try it.

Reply:
@johnpmitsch I had actually played around with other data shapes like `{4: {expandedColumn: 5 } }` but deleted the comment because I liked it less than what you have.  The only other thing I can think of is having `rows` have an additional level, so it could store both `rowData` and `rowMetadata` on the same object.  But haven't thought out completely how that would work.